rules_daml/daml-build.bzl

```diff
@@ -0,0 +1,112 @@
+# Copyright (c) 2020 The DAML Authors. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+def file_of_target(k):
+    [file] = k.files.to_list()
+    return file
+
+_damlc = attr.label(
+    allow_single_file = True,
+    default = Label("//compiler/damlc"),
+    executable = True,
+    cfg = "host",
+)
+
+def _daml_build_impl(ctx):
+    name = ctx.label.name
+    config = ctx.file.config
+    srcs = ctx.files.srcs
+    dd = ctx.attr.dar_dict
+    damlc = ctx.file._damlc
+    project_root = config.dirname
+    input_dars = [file_of_target(k) for k in dd.keys()]
+    output_dar = ctx.actions.declare_file(name + ".dar")
+    ctx.actions.run_shell(
+        tools = depset([damlc]),
+        inputs = depset([config] + srcs + input_dars),
+        outputs = [output_dar],
+        progress_message = "Building DAML project %s" % name,
+        command = """
```

Comment:
I think there are two usecases for a `daml_build` rule:

1. Documentation for upgrading, packaging, …. In this case the `daml.yaml` is something displayed in the documentation and we want to test against that exact `daml.yaml` (see my comment above on the change to the `daml.yaml`). I think in that case we might be able to get away with something simpler: We can not shuffle sources around and instead just go to the directory of the `daml.yaml` and call `daml build`. If the files are arranged in the right locations I believe this should work out. The one thing that won’t work is finding the `dar` files. I don’t really have a great solution for that. idea. I don’t expect this usecase to ever become particularly common. ATM the upgrading docs are the only thing that would make use of this and I wonder if we might just be better off keeping the `sh_test` there to make sure we reproduce exactly what our documentation says. Curious what @aherrmann-da thinks about this.

2. The other usecase is basically every single call to `daml_compile` in our codebase atm and things like the genrule for triggers or DAML script which cannot use `daml_compile`. I think in this case, we don’t want to specify a `daml.yaml` at all. Instead if probably makes sense for the Bazel rule to accept a list of DAR files as dependencies, a list of source files, a name, and maybe a `strip_prefix` field and then the rule will generate the `daml.yaml` automagically.

I think the current rule is in a bit of an awkward middle ground where it doesn’t quite work for documentation since it requires modifying the `daml.yaml` but it’s also rather awkward to use outside of documentation since there it would be much more convenient to specify the `daml.yaml` directly in the rule. Given that 2 is much more common, I think it would make more sense to at least initially focus the rule on that and leave documentation like it is for now.

Reply:
I think part of what is missing is a principled way of locating one DAML project from another with a `daml.yaml` file

Reply:
I'm not going to suggest this rule is perfect, but I think it's a definite step forward from the use of `sh_test`.

Reply:
> I think part of what is missing is a principled way of locating one DAML project from another within a `daml.yaml` file.

The problem with that is that in general you might only have the DAR and nothing else. Pointing to the sources doesn’t really make sense. It is crucial that you use the DAR as the input and don’t recompile the sources for `data-dependencies`.

> I'm not going to suggest this rule is perfect, but I think it's a definite step forward from the use of `sh_test`.

The `sh_test` tests exactly what we want users to run. With the change to the `daml.yaml` this won’t run as all unless you explicitly copy the DAR around which isn’t explained in the documentation and I don’t think we want to explain it since it’s just another step with no benefit. So while I do agree that it’s nicer in terms of the code in the `BUILD.bazel` file, it’s worse in terms of the UX for someone reading the docs which is the important part here. That’s why I think focusing on 2 where we can completely ignore those concerns might be more sensible.

Reply:
> 1. The one thing that won’t work is finding the `dar` files. I don’t really have a great solution for that.

Could this be done with `--package-db` flags? `daml_compile` (or a dedicated `daml_package`) could generate one, and we could pass `--package-db` flags to `damlc` for dependencies.

Reply:
> I think there are two usecases for a `daml_build` rule
> [...]
> 2\. The other usecase is basically every single call to `daml_compile` in our codebase atm and things like the genrule for triggers or DAML script which cannot use `daml_compile`. I think in this case, we don’t want to specify a `daml.yaml` at all. Instead if probably makes sense for the Bazel rule to accept a list of DAR files as dependencies, a list of source files, a name, and maybe a `strip_prefix` field and then the rule will generate the `daml.yaml` automagically.

I've been wondering about that. What's unfortunate about this approach is that we'd have to keep the available metadata fields in `daml.yaml` in sync with the corresponding Bazel rule's available attributes. If we could figure out how to make `daml.yaml` work within Bazel, that would seem preferable.

If it's only the DAR paths, maybe we could have a search paths flag to `damlc`? Something akin to `-I`/`-L`.

Reply:
> The problem with that is that in general you might only have the DAR and nothing else. Pointing to the sources doesn’t really make sense. It is crucial that you use the DAR as the input and don’t recompile the sources for data-dependencies.

This of course is true for real uses of old, mustn't-be-changed dar files.

But the rule under discussion is intended to check the DAML of our documentation example is correct. So we do want to have a proper dependency on v1 form v2. So if someone edits the DAML in v1 to break v2, the test will discover this.

Reply:
> > I've been wondering about that. What's unfortunate about this approach is that we'd have to keep the available metadata fields in `daml.yaml` in sync with the corresponding Bazel rule's available attributes.
> 
> I would just remove the `daml.yaml` file completely for those cases so there’s nothing to keep in sync.

I don't mean the values, I mean the fields themselves. E.g. the `daml_compile` rule may need a `exposed-modules` attribute, or a `build-options` attribute. It's not too bad, but it's repetition. For the cases we have so far, it's probably fine.

Reply:
Ah I see. Yeah that’s slightly annoying. Anyway that’s a discussion for a separate PR.